File: RiceCooker.ts
import { question } from 'readline-sync';

export class RiceCooker {
  public cookedRice: boolean = false;

  private async cookItem(ingredient: string, successMessage: string): Promise<boolean> {
    try {
      console.log(`Ajout de ${ingredient}...`);
      await this.delay(1000);
      this.beep();
      console.log(`${successMessage} terminée !`);
      return true;
    } catch (error) {
      console.error(`Une erreur s'est produite lors de la cuisson de ${ingredient}: ${error}`);
      return false;
    }
  }
  

  public async cookRice(): Promise<boolean> {
    return this.cookItem("du riz", "Cuisson du riz");
  }

  public async keepWarm(): Promise<boolean> {
    return this.cookItem("de la nourriture", "Maintien au chaud de la nourriture");
  }

  public async steamCook(): Promise<boolean> {
    return this.cookItem("de la nourriture pour la cuisson à la vapeur", "Cuisson à la vapeur");
  }

  public async cookSoup(): Promise<boolean> {
    const waterAdded = await this.cookItem("de l'eau", "Cuisson de la soupe");
    return waterAdded;
  }

  public async cookDessert(): Promise<boolean> {
    return this.cookItem("des ingrédients pour le dessert", "Cuisson du dessert");
  }

  public beep(): void {
    console.log("BEEP BEEP BEEP - Cuisson terminée !");
  }

  public async start(): Promise<void> {
    let choice: number | null = null;

    do {
      console.log("\nQue souhaitez-vous faire ?");
      console.log("1. Cuire du riz");
      console.log("2. Maintenir au chaud");
      console.log("3. Cuisson à la vapeur");
      console.log("4. Cuisson de la soupe");
      console.log("5. Cuisson du dessert");
      console.log("6. Quitter");

      const input = question('Choisissez une option : ');

      choice = Number(input);

      if (isNaN(choice)) {
        console.log("Choix non valide. Veuillez choisir une option valide");
        choice = 0;
      }

      switch (choice) {
        case 1:
          await this.cookRice();
          break;
        case 2:
          await this.keepWarm();
          break;
        case 3:
          await this.steamCook();
          break;
        case 4:
          await this.cookSoup();
          break;
        case 5:
          await this.cookDessert();
          break;
        case 6:
          console.log("Au revoir !");
          break;
      }
    } while (choice !== 6);
  }

  public async delay(ms: number): Promise<void> {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

if (require.main === module) {
  const riceCooker = new RiceCooker();
  riceCooker.start();
}

File: rice_cooker.rb
# frozen_string_literal: true

# Simulates the functionality of a rice cooker with various cooking options.
class RiceCooker
    def initialize
      @cooked_rice = false
    end
  
    def cook_item(ingredient, success_message)
      puts "Ajout de #{ingredient}..."
      delay(1000)
      beep
      puts "#{success_message} terminée !"
      true
    rescue StandardError => e
      puts "Une erreur s'est produite lors de la cuisson de #{ingredient}: #{e.message}"
      false
    end
  
    def cook_rice
      cook_item('du riz', 'Cuisson du riz')
    end
  
    def keep_warm
      cook_item('de la nourriture', 'Maintien au chaud de la nourriture')
    end
  
    def steam_cook
      cook_item('de la nourriture pour la cuisson à la vapeur', 'Cuisson à la vapeur')
    end
  
    def cook_soup
      cook_item("de l'eau", 'Cuisson de la soupe')
    end
  
    def cook_dessert
      cook_item('des ingrédients pour le dessert', 'Cuisson du dessert')
    end
  
    def beep
      puts 'BEEP BEEP BEEP - Cuisson terminée !'
    end
  
    def start
      loop do
        display_options
        choice = user_choice
  
        handle_choice(choice)
        break if choice == 6
      end
    end
  
    private
  
    def display_options
      puts "\nQue souhaitez-vous faire ?"
      puts '1. Cuire du riz'
      puts '2. Maintenir au chaud'
      puts '3. Cuisson à la vapeur'
      puts '4. Cuisson de la soupe'
      puts '5. Cuisson du dessert'
      puts '6. Quitter'
      print 'Choisissez une option : '
    end
  
    def user_choice
      input = gets.chomp
      input.to_i.zero? && input != '0' ? 0 : input.to_i
    end
  
    def handle_choice(choice)
      case choice
      when 1 then cook_rice
      when 2 then keep_warm
      when 3 then steam_cook
      when 4 then cook_soup
      when 5 then cook_dessert
      when 6 then quit
      else invalid_choice
      end
    end
  
    def quit
      puts 'Au revoir !'
      break
    end
  
    def invalid_choice
      puts 'Choix non valide. Veuillez choisir une option valide'
    end
  
    def delay(milliseconds)
      sleep(milliseconds / 1000.0)
    end
  end
  
  if __FILE__ == $PROGRAM_NAME
    rice_cooker = RiceCooker.new
    rice_cooker.start
  end
  